gnrc_netdev2: allow for reception of zero-length payload

miri64 opened the pull request:
Fixes #5535

aeneby commented:
Thanks for looking into this. However, although the crash is gone, zero-length packets still don't seem to get output by `pktdump`. Perhaps this is a bug in `pktdump` - I assume that the zero-length packets are still making their way through the entire stack?

I also notice an inconsistency in the debug output between `gnrc_netdev2_ieee802154` and `gnrc_netdev2_eth` devices - on receipt of an empty data packet, the eth device prints:

`gnrc_netdev2_eth: received packet from 5a:58:9c:29:23:f1 of length 14`

Whereas the corresponding 802154 device prints:

`_recv_ieee802154: received packet from ea:e3 of length 0`

miri64 commented:
> Thanks for looking into this. However, although the crash is gone, zero-length packets still don't seem to get output by pktdump. Perhaps this is a bug in pktdump - I assume that the zero-length packets are still making their way through the entire stack?

Ah, forgot to mention this: no that is in fact "wanted". Since the payload is missing the packet just has `NETIF` as type. One needs to register for that type too. I decided against that, because 0 length is a special use-case. About the inconsistent debug output: the IEEE 802.15.4 output seems to add the header while the Ethernet module doesn't.

aeneby commented:
>I decided against that, because 0 length is a special use-case.

I disagree, there's nothing special about a data packet with zero-length payload. There are other types of "dataless" packets too, such as acknowledgement response packets, which therefore don't show up in `pktdump` either. I don't see any reason for these to be ignored.

>About the inconsistent debug output: the IEEE 802.15.4 output seems to add the header while the Ethernet module doesn't.

Yep, suspected as much. Would it make sense to have some consistency across link-layer types though, do you think?

miri64 commented:
>> I decided against that, because 0 length is a special use-case.
> I disagree, there's nothing special about a data packet with zero-length payload. There are other types of "dataless" packets too, such as acknowledgement response packets, which therefore don't show up in pktdump either. I don't see any reason for these to be ignored.

Sorry, I wasn't clear about that: for `pktdump` this *is* a special case (since you have to register a whole new type), there is nothing to have an opinion about. We could register `NETIF` with `pkgdump`, but I tend not to, to keep the simplicity of the example high.

> Yep, suspected as much. Would it make sense to have some consistency across link-layer types though, do you think?

This PR is not about consistency of debug outputs (it isn't consistent for most of the system), but about fixing a crash issue. So I would prefer to do this in a separate PR as soon as this one is merged.

aeneby commented:
>We could register NETIF with pkgdump, but I tend not to, to keep the simplicity of the example high.

Okay, so this restriction only applies the default example, and is not an intrinsic limitation of pkgdump itself? Admittedly I'm not familiar with the concept of registering types; I'll have to read up on it.

>I would prefer to do this in a separate PR as soon as this one is merged.

Roger that.

miri64 commented:
> Okay, so this restriction only applies the default example, and is not an intrinsic limitation of pkgdump itself? Admittedly I'm not familiar with the concept of registering types; I'll have to read up on it.

Yes. The type of a packet is defined by the type of its first packet snip. Since a L2 packet normally contains a payload snip and a `netif_hdr_t` snip, `pktdump` just registers for the payload (`UNDEF` in this case, because no 6LoWPAN or other network layer protocol is present in the default example). Mind that in reception are in ["reversed" order](http://doc.riot-os.org/group__net__gnrc__pkt.html#ga278e783e56a5ee6f1bd7b81077ed82a7). When no payload is present, the packet just has the `netif_hdr_t` snip, which of course is of type `NETIF`, so the overall packet's type is also `NETIF`. Since `pktdump` in the default example is not registered to that, they are ignored.

aeneby commented:
I see, understood. Thanks for the clarification.

aeneby commented:
Sorry, One More Thing™ - I can overflow the packet buffer in the default example by repeatedly spamming a node with empty data packets. So to reproduce, set up a two-node native port topology, and spam one node from the other:

`txtsnd 4 2a:58:38:e2:6a:00 ""`

After 15-20 packets the receiving node will start complaining:

`_recv_ethernet_packet: cannot allocate pktsnip.`

I also noticed this using 802.15.4 nodes.

miri64 commented:
Ok, will look into that.

miri64 commented:
Rebased to current master to include #5534. And fixed the issue (forgot to remove the original L2 header in the case of empty payload). As a bonus I unified the debug output for packet reception in IEEE 802.15.4 and Ethernet. Both now print the full received packet including original header.

miri64 commented:
Ah wait -.- should have tested... now I've got a segfault

miri64 commented:
Fixed.

aeneby commented:
Nice. Works fine for me now.

I still can't get `pktdump` to dump `NETTYPE_NETIF` (-1) packets though, since `gnrc_nettype_register()` [won't allow](http://riot-os.org/api/group__net__gnrc__netreg.html#gafb6a9d0b840aae07c3c1617932835ef1) registering types < `NETTYPE_UNDEF` (0). Is there a reason for this?

miri64 commented:
Probably legacy comstraints `NETIF` was introduced later (that's why it's -1).

aeneby commented:
@kaspar030 could we get this put on the 2016.07 milestone, perhaps? It's quite an important PR since without it it's quite trivial to DoS RIOT nodes.

miri64 commented:
As discussed in #5542: allowing for 0-length packet snips might be the cleaner solution in the end. I would like to wait for some discussion results there before closing this one and opening a new one to adapt the packet buffer for that case.

kYc0o commented:
So, the discussion took place and we can close this?

miri64 commented:
Leave it open as a reminder for me for now please. Will do this post release.

kaspar030 commented on sys/net/gnrc/link_layer/netdev2/gnrc_netdev2_eth.c:
```diff
@@ -99,9 +102,18 @@ static gnrc_pktsnip_t *_recv(gnrc_netdev2_t *gnrc_netdev2)
-        LL_APPEND(pkt, netif_hdr);
+        /* if pkt has payload */
+        if (pkt != eth_hdr) {
+            /* replace IEEE 802.15.4 header with generic netif header */
```
802.15.4?

miri64 replied:
This PR needs rework anyway. So thanks, but this line will probably disappear ;-)

miri64 commented:
Replaced by #5694.
